lib/Dialect/HW/HWOps.cpp

```diff
@@ -1956,6 +1964,23 @@ void StructInjectOp::print(OpAsmPrinter &printer) {
   printer << " : " << getInput().getType();
 }
 
+LogicalResult StructInjectOp::canonicalize(StructInjectOp op,
+                                           PatternRewriter &rewriter) {
+  auto inputOp = op.getInput().getDefiningOp();
+
+  // b = inject(inject(x["field"], v0)["field"], v1) => inject(x["field"], v1)
+  if (auto structInject = dyn_cast_or_null<StructInjectOp>(inputOp)) {
```

Comment:
In a graph region, couldn't this cause an infinite loop?
`%a = inject(%a["field], v)`

Reply:
`%a = inject((inject(%a["field], v))["field], v) would canonicalize to `%a = inject(%a["field], v)`, so there is a struct reduction step. If we only see `%a = inject(%a["field], v)` then nothing happens.

Reply:
I agree with Lenharth, this IR causes an infinite loop.
```
hw.module @ExtractOfInject(%a: !hw.struct<a: i1>, %v: i1) -> (result: i1) {
  %b = hw.struct_inject %b["a"], %v : !hw.struct<a: i1>
  %c = hw.struct_extract %b["a"] : !hw.struct<a: i1>
  // CHECK: hw.output %v
  hw.output %c : i1
}
```

Reply:
I stand corrected. Fixed the cycle.